blob/service.go

```diff
@@ -347,127 +354,204 @@ func (s *Service) retrieve(
+				// now that we found the requested blob, we will create
+				// its inclusion proof.
+				inclusiveBlobStartRowIndex := inclusiveNamespaceStartRowIndex + currentShareIndex/squareSize
+				exclusiveBlobEndRowIndex := inclusiveNamespaceStartRowIndex + endShareIndex/squareSize
+				if (currentShareIndex+blobLen)%squareSize != 0 {
+					// if the row is not complete with the blob shares,
+					// then we increment the exclusive blob end row index
+					// so that it's exclusive.
+					exclusiveBlobEndRowIndex++
+				}
 
-			index += len(appShares) - len(shrs)
-			appShares = shrs
-			sharesParser.reset()
+				// create the row roots to data root inclusion proof
+				rowProofs := proveRowRootsToDataRoot(
+					append(header.DAH.RowRoots, header.DAH.ColumnRoots...),
+					inclusiveBlobStartRowIndex,
+					exclusiveBlobEndRowIndex,
+				)
+				rowRoots := make([]bytes.HexBytes, exclusiveBlobEndRowIndex-inclusiveBlobStartRowIndex)
+				for index, rowRoot := range header.DAH.RowRoots[inclusiveBlobStartRowIndex:exclusiveBlobEndRowIndex] {
+					rowRoots[index] = rowRoot
+				}
 
-			if !wasEmpty {
-				// remove proofs for prev rows if verified blob spans multiple rows
-				proofs = proofs[len(proofs)-1:]
-			}
-		}
+				// create the share to row root proofs
+				shareToRowRootProofs, _, err := pkgproof.CreateShareToRowRootProofs(
+					squareSize,
+					//nolint:lll
+					rowsWithParityShares[inclusiveBlobStartRowIndex-inclusiveNamespaceStartRowIndex:exclusiveBlobEndRowIndex-inclusiveNamespaceStartRowIndex],
+					header.DAH.RowRoots[inclusiveBlobStartRowIndex:exclusiveBlobEndRowIndex],
+					currentShareIndex%squareSize,
+					(endShareIndex-1)%squareSize,
+				)
+				if err != nil {
+					return nil, nil, err
+				}
 
-		rowIndex++
-		if sharesParser.isEmpty() {
-			proofs = nil
+				proof := Proof{
+					ShareToRowRootProof: shareToRowRootProofs,
+					RowProof: core.RowProof{
+						RowRoots: rowRoots,
+						Proofs:   rowProofs,
+						StartRow: uint32(inclusiveBlobStartRowIndex),
+						EndRow:   uint32(exclusiveBlobEndRowIndex),
+					},
+				}
+				return blob, &proof, nil
+			}
+			sharesParser.reset()
+			currentShareIndex += blobLen
+		} else {
+			// this is a continuation of a previous blob
+			// we can skip
+			currentShareIndex++
 		}
 	}
+	return nil, nil, ErrBlobNotFound
+}
 
-	err = ErrBlobNotFound
-	for _, sh := range appShares {
-		ok, err := sh.IsPadding()
-		if err != nil {
-			return nil, nil, err
-		}
-		if !ok {
-			err = fmt.Errorf("incomplete blob with the "+
-				"namespace: %s detected at %d: %w", namespace.String(), height, err)
-			log.Error(err)
-		}
-	}
-	return nil, nil, err
+// proveRowRootsToDataRoot creates a set of binary merkle proofs for all the
+// roots defined by the range [start, end).
+func proveRowRootsToDataRoot(roots [][]byte, start, end int) []*merkle.Proof {
+	_, proofs := merkle.ProofsFromByteSlices(roots)
+	return proofs[start:end]
```

Comment:
Can we move this function to the blob.go, where Proof is defined?

Reply:
[11d425c](https://github.com/celestiaorg/celestia-node/pull/3610/commits/11d425c3fe02293c3229f14bad985d6a5a075f56)